test/dist_autograd_test.py

```diff
@@ -187,6 +272,101 @@ def test_autograd_functions(self):
 
+    @dist_init
+    def test_graph_for_builtin_call(self):
+        self._test_graph(torch.add)
+
+    @dist_init
+    def test_graph_for_python_call(self):
+        self._test_graph(my_py_add)
+
+    # 3-layer nested calls
+    @dist_init
+    def test_graph_for_py_nested_call(self):
+        dst_rank = (self.rank + 1) % self.world_size
+        with dist_autograd.context() as context_id:
+            t1 = torch.ones(3, 3, requires_grad=True)
+            t2 = torch.zeros(3, 3, requires_grad=True)
+            nest_dst_rank = (dst_rank + 1) % self.world_size
+            ret = rpc.rpc_sync("worker{}".format(dst_rank),
+                               my_py_nested_call, args=(t1, t2, dst_rank, self.world_size, 1))
+            for rd in RankDistance:
+                rpc.rpc_sync("worker{}".format((self.rank + rd) % self.world_size),
+                             _set_rpc_done, args=(context_id, rd))
+
+            # For self.rank, it has 4 graphs to verify
+            # One is for current context id when this rank send first rpc call.
+            # Second one is for prev context id when this rank make 1st nested
+            # call.
+            # Third one is for prev prev context id when this rank make
+            # 2nd nested call.
+            # Last one is for prev prev prev context id when this rank
+            # execute the torch.add() operator.
+
+            # Verify first graph for current context id.
+            ctx = dist_autograd._current_context()
+            self.assertEqual(context_id, ctx._context_id())
+            send_functions = ctx._send_functions()
+            self.assertEqual(1, len(send_functions))
+            recv_functions = ctx._recv_functions()
+            self.assertEqual(1, len(recv_functions))
+            self._verify_graph_for_first_rpc_call(list(send_functions.values())[0],
+                                                  list(recv_functions.values())[0],
+                                                  t1, t2, ret)
+
+            # Verify second graph for 1st nested call.
+            self._check_rpc_done(RankDistance.PREV)
+            ctx = dist_autograd._retrieve_context(ctx_ids[RankDistance.PREV])
+            self._verify_graph_for_nested_rpc_call(ctx)
+
+            # Verify third graph for 2nd nested call.
+            self._check_rpc_done(RankDistance.PREV_PREV)
+            ctx = dist_autograd._retrieve_context(ctx_ids[RankDistance.PREV_PREV])
+            self._verify_graph_for_nested_rpc_call(ctx)
+
+            # verify last graph for rpc call execution.
+            self._check_rpc_done(RankDistance.PREV_PREV_PREV)
+            ctx = dist_autograd._retrieve_context(ctx_ids[RankDistance.PREV_PREV_PREV])
+            send_functions = ctx._send_functions()
+            self.assertEqual(1, len(send_functions))
+            self._verify_graph_for_rpc_call_exec(list(send_functions.values())[0])
+
+    # Rank0->Rank1->Rank0
+    @dist_init
```

Comment:
Can we also add a test where none of the tensors require grad and verify that we don't attach send/recv functions anywhere?

Reply:
yes, and need to verify no prev context is passed over when there is no tensors requiring grads.